### PR TITLE
feat(core): add keys and key-sets entity

### PR DESCRIPTION
**Pull request by jschmid1**
Supersedes: https://github.com/Kong/kong/pull/9677

This is ready for a first round of reviews.

### Summary

This PR adds support for managing asymmetric keys in various formats (JWK, PEM). It adds two new entities

* /keys
* /key-sets

that accept the mentioned formats.

### Full changelog

* add new core entities keys and key-sets
* adds daos for keys and key-sets entities

Missing things:
- [x] c* support ~or skip tests for it~
- [x] proper uniqueness of keys within a set
- [ ] docs (queued after initial review/approval)

Signed-off-by: Joshua Schmid <jaiks@posteo.de>

### CODE REVIEW

File: kong/db/migrations/core/017_300_to_310.lua [hunk 1]
```diff
@@ -8,13 +8,115 @@ return {
...
+          -- do nothing, accept existing state
+        end$$;
+
+        create table if not exists "keys" (
+          "id"           uuid                       primary key,
+          "set_id"       uuid                       REFERENCES "key_sets" ("id") on delete cascade,
+          "name"         text                       unique,
+          "cache_key"    text                       unique,
+          "ws_id"        UUID                       REFERENCES "workspaces" ("id"),
+          "kid"          text,                      unique ("kid", "set_id"),
```
Comment (bungle) [inline kong/db/migrations/core/017_300_to_310.lua:47]: perhaps move the unique constraint at the bottom.

File: kong/db/schema/entities/keys.lua [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        type     = "string",
+        required = false,
+        unique   = true,
+      },
+    },
+    {
+      kid = {
+        type     = "string",
+        required = true,
+        unique   = true,
```
Comment (bungle) [inline kong/db/schema/entities/keys.lua:44]: maybe this should not be unique

File: kong/db/schema/typedefs.lua [hunk 3]
```diff
@@ -573,6 +573,227 @@ typedefs.semantic_version = Schema.define {
...
+        type = "string",
+        required = false,
+        referenceable = true,
+        encrypted = true
+      },
+    },
+    {
+      public_key = {
+        type = "string",
+        required = false,
```
Comment (bungle) [inline kong/db/schema/typedefs.lua:624]: maybe this could be referenceable too.

File: kong/db/schema/entities/keys.lua [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        if type(entity.jwk) == "string" then
+          if kong.vault.is_reference(entity.jwk) then
+            -- can't validate a reference
+            return true
+          end
+          -- validate against the typedef.jwk
+          local schema = Schema.new(typedefs.jwk)
+          if not schema then
+            return nil, "could not load jwk schema"
+          end
```
Comment (kikito) [inline kong/db/schema/entities/keys.lua:91]: typedefs.jwk should already be an instance of Schema, so invoking Schema.new again shouldn't be needed here (untested):

```suggestion
          local schema = typedefs.jwk
```

File: kong/db/schema/entities/keys.lua [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            return nil, "kid in jwk.kid must be equal to keys.kid"
+          end
+
+          -- running customer_validator
+          local ok, val_err = typedefs.jwk.custom_validator(entity.jwk)
+          if not ok or val_err then
+            return nil, "could not load jwk"
+          end
+          -- FIXME: this does not execute the `custom_validator` part.
+          --        how to do that without loading that manually as seen above
```
Comment (kikito) [inline kong/db/schema/entities/keys.lua:110]: By reading the schema/init.lua file: `schema:validate` should reach it by doing `run_entity_checks` -> `run_check` ->  The `custom_validator` function should be part of `self.entity_checks`.

It is possible that this has not happened on your tests if the object is not valid on a field level - the individual field validators are run, if an error is found on the fields then the entity validator is not run.

File: kong/db/schema/typedefs.lua [hunk 6]
```diff
@@ -573,6 +573,227 @@ typedefs.semantic_version = Schema.define {
   },
 }
 
+local function validate_jwk(key)
+  -- unless it's a reference
```
Comment (kikito) [inline kong/db/schema/typedefs.lua:577]: This comment is a bit "concatenate name and surname"
```suggestion
```

File: kong/db/schema/typedefs.lua [hunk 7]
```diff
@@ -573,6 +573,227 @@ typedefs.semantic_version = Schema.define {
...
+  if kong.vault.is_reference(key) then
+    return true
+  end
+
+  local pk, err = openssl_pkey.new(key, { format = "JWK" })
+  if not pk or err then
+    return false, "could not load JWK"
```
Comment (kikito) [inline kong/db/schema/typedefs.lua:584]: Should we return `err` here?
```suggestion
    return false, "could not load JWK " .. (err or "")
```

File: kong/db/schema/typedefs.lua [hunk 8]
```diff
@@ -573,6 +573,227 @@ typedefs.semantic_version = Schema.define {
...
+    return false, "could not load public key"
+  end
+
+  -- unless it's a reference
+  if kong.vault.is_reference(private_key) then
+    return true
+  end
+  local ppk, perr = openssl_pkey.new(private_key, { format = "PEM" })
+  if not ppk or perr then
+    return false, "could not load private key"
```
Comment (kikito) [inline kong/db/schema/typedefs.lua:604]: ditto for perr

```suggestion
    return false, "could not load private key " .. (perr or "")
```

File: kong/constants.lua [hunk 9]
```diff
@@ -82,6 +82,15 @@ for p,_ in pairs(protocols_with_subsystem) do
 end
 table.sort(protocols)
 
+local key_formats_map = {
+    ["jwk"] = true,
+    ["pem"] = true,
+}
```
Comment (flrgh) [inline kong/constants.lua:88]: indentation fixage:

```suggestion
local key_formats_map = {
  ["jwk"] = true,
  ["pem"] = true,
}
```

File: kong/constants.lua [hunk 10]
```diff
@@ -82,6 +82,15 @@ for p,_ in pairs(protocols_with_subsystem) do
...
+local key_formats = {}
+for k,_ in pairs(key_formats_map) do
```
Comment (flrgh) [inline kong/constants.lua:90]:
```suggestion
for k in pairs(key_formats_map) do
```

File: kong/db/dao/keys.lua [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+  return fmt("keys:%s:%s", tostring(kid), set_name)
+end
+
+-- load to lua-resty-openssl pkey module
+local function _load_pkey(key, part)
+  local pk, err
+  local part_to_field = {
+    ["public"] = "public_key",
+    ["private"] = "private_key",
+  }
```
Comment (flrgh) [inline kong/db/dao/keys.lua:110]: Can this table creation be moved up to the outer scope? Or, given that there are only 2 items in this alias/lookup table, can it be omitted in favor of just changing the API to `_load_pkey("my-key", "public_key")`?

File: kong/db/migrations/core/017_300_to_310.lua [hunk 12]
```diff
@@ -8,13 +8,115 @@ return {
             -- Do nothing, accept existing state
             END;
         $$;
+
+        create table if not exists "key_sets" (
```
Comment (flrgh) [inline kong/db/migrations/core/017_300_to_310.lua:12]: It's pretty minor, but would you mind adjusting the case of your SQL to match the rest of the migrations (i.e. `create table` => `CREATE TABLE`)?

**Review by flrgh — COMMENTED** [review]
Looks very solid! All high level items have been well-covered by other reviewers, so for the most part my comments are all low-hanging fruit (formatting/consistency, code style, etc).

File: kong/db/dao/key_sets.lua [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+-- subject to the terms of the Kong Master Software License Agreement found
+-- at https://konghq.com/enterprisesoftwarelicense/.
+-- [ END OF LICENSE 0867164ffc95e54f04670b5169c09574bdbd9bba ]
+
+local key_sets = {}
+
+
+function key_sets:truncate()
+  return self.super.truncate(self)
+end
```
[the remark this answers is not in the document]
Comment (jschmid1) [reply kong/db/dao/key_sets.lua:13]: A matter of taste I guess. Some of the methods are generated based on presence of certain entity fields. Explicitly defining the methods, even if these are just a pass-through, helps with understanding the functionality that this dao provides.

File: kong/db/schema/entities/keys.lua (hunk 4 above)
Comment (jschmid1) [reply kong/db/schema/entities/keys.lua:91]: `typedefs.jwk` is a schema _definition_ but not a Schema instance.

File: kong/db/dao/keys.lua (hunk 11 above)
Comment (jschmid1) [reply kong/db/dao/keys.lua:110]: This list is expected to grow to at least len+1. I'll move it up the scope 👍🏻

File: kong/db/migrations/core/017_300_to_310.lua (hunk 12 above)
Comment (jschmid1) [reply kong/db/migrations/core/017_300_to_310.lua:12]: uppercase'd all-the-things

File: kong/db/schema/entities/keys.lua (hunk 5 above)
Comment (jschmid1) [reply kong/db/schema/entities/keys.lua:110]: Right,[ it _should_ do that.](https://github.com/Kong/kong/blob/master/kong/db/schema/init.lua#L981) I'll leave the deeper analysis for later as we have a valid workaround.

**Review by bungle — APPROVED** [review]
Let's move on with this, thanks @jschmid1.

**Pull request merged**